cmd/juju/model/defaultscommand.go

```diff
@@ -97,120 +127,425 @@ func (c *defaultsCommand) SetFlags(f *gnuflag.FlagSet) {
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	keyCounter := map[string]int{}
+	c.values = make(attributes)
+	for k, v := range options {
+		if k == config.AgentVersionKey {
+			return errors.Errorf(`%q must be set via "upgrade-juju"`, config.AgentVersionKey)
+		}
+		keyCounter[k]++
+		c.values[k] = v
+	}
+	for _, resetKey := range c.resetKeys {
+		keyCounter[resetKey]++
+		if keyCounter[resetKey] > 1 {
+			return errors.Errorf(`key %q specified more than once`, resetKey)
+		}
 	}
-	c.action = c.getDefaults
 	return nil
 }
 
-// getAPI sets the api on the command. This allows passing in a test
-// ModelDefaultsAPI implementation.
-func (c *defaultsCommand) getAPI() (defaultsCommandAPI, error) {
-	if c.api != nil {
-		return c.api, nil
+// handleSetArgs parses args for setting defaults.
+func (c *defaultsCommand) handleSetArgs(args []string) error {
+	argZeroKeyOnly := !strings.Contains(args[0], "=")
+	// If an invalid region was specified, the first positional arg won't have
+	// an "=". If we see one here we know it is invalid.
+	switch {
+	case argZeroKeyOnly && c.regionName == "":
+		return errors.Errorf("invalid region specified: %q", args[0])
+	case argZeroKeyOnly && c.regionName != "":
+		return errors.New("cannot set and retrieve default values simultaneously")
+	default:
+		if err := c.parseSetKeys(args); err != nil {
+			return errors.Trace(err)
+		}
+		c.action = c.setDefaults
+		return nil
 	}
+}
 
-	api, err := c.NewAPIRoot()
-	if err != nil {
-		return nil, errors.Annotate(err, "opening API connection")
+// handleZeroArgs determines whether we are doing a reset only and if any
+// additional arguments are valid -- or why they are not.
+func (c *defaultsCommand) handleZeroArgs() error {
+	resetSpecified := c.resetKeys != nil
+	regionSpecified := c.regionName != ""
+
+	switch {
+	case regionSpecified && !resetSpecified:
+		// It doesn't make sense to specify a region unless we are resetting
+		// values here.
+		return errors.New("specifying a region when retrieving defaults is invalid")
+	case regionSpecified && resetSpecified:
+		// We can reset for a region, c.action stays nil. We want to do reset
+		// only.
+		return nil
+	case !regionSpecified && resetSpecified:
+		// Again we want to reset, this is for resetting the controller as
+		// there is no region specified. So c.action remains nil.
+		return nil
+		// We short circuited at the begining of c.Init so we should never
+		// reach a case where "!regionSpecified && !resetSpecified"
+	default:
+		// This should be unreachable.
+		return errors.New("unexpected input, please file a bug")
 	}
-	client := modelmanager.NewClient(api)
+}
 
-	return client, nil
+// handleOneArg handles the case where we have one positional arg after
+// processing for a region and the reset flag after processing for a region and
+// the reset flagg.
+func (c *defaultsCommand) handleOneArg(arg string) error {
+	resetSpecified := c.resetKeys != nil
+	regionSpecified := c.regionName != ""
+
```

Comment:
Maybe we can revisit this after we try out a couple snaps to find out if this will stick or if we'll put regions behind a flag.  That said, I agree that just the code like that reads better. But documenting the conditions and their implications doesn't read as well, IMO. If region goes behind a flag I think parsing would be simpler and a bunch of ugly code could simply disappear :)

```Go
      if regionSpecified {
         if !resetSpecified {
             // It doesn't make sense to specify a region unless we are
             // resetting values here.
             return errors.New("specifying a region when retrieving defaults for     a setting is invalid")
         }
         // If a region was specified and reset was specified, we shouldn't have
         // an extra arg. If it had an "=" in it, we should have handled it
         // already.
         return errors.New("cannot retrieve defaults for a key and reset args at     the same time")
     }
     if resetSpecified {
         // It makes no sense to supply a positional arg that isn't a region if
         // we are resetting a region, so we must have gotten an invalid region.
         return errors.Errorf("invalid region specified: %q", arg)
     }
     // We can retrieve a value.
     c.key = arg
     c.action = c.getDefaults
     return nil
```